src/solvers/refinement/string_refinement.cpp

```diff
@@ -0,0 +1,826 @@
+        if(seen_instances.insert(instance).second)
+          add_lemma(instance);
+        else
+          debug() << "instance already seen" << eom;
+      }
+    }
+
+    return false;
+  }
+}
+
+
+std::map<exprt, int> string_refinementt::map_of_sum(const exprt &f)
+{
+  // number of time the element should be added (can be negative)
+  std::map<exprt, int> elems;
+
+  std::vector<std::pair<exprt, bool> > to_process;
+  to_process.push_back(std::make_pair(f, true));
+
+  while(!to_process.empty())
+  {
+    exprt cur=to_process.back().first;
+    bool positive=to_process.back().second;
+    to_process.pop_back();
+    if(cur.id()==ID_plus)
+    {
+      to_process.push_back(std::make_pair(cur.op1(), positive));
+      to_process.push_back(std::make_pair(cur.op0(), positive));
+    }
+    else if(cur.id()==ID_minus)
+    {
+      to_process.push_back(std::make_pair(cur.op1(), !positive));
+      to_process.push_back(std::make_pair(cur.op0(), positive));
+    }
+    else if(cur.id()==ID_unary_minus)
+    {
+      to_process.push_back(std::make_pair(cur.op0(), !positive));
+    }
+    else
+    {
+      if(positive)
+        elems[cur]=elems[cur]+1;
+      else
+        elems[cur]=elems[cur]-1;
+    }
+  }
+  return elems;
+}
+
+
+exprt string_refinementt::sum_of_map(std::map<exprt, int> & m, bool negated)
+{
+  exprt sum=refined_string_typet::index_of_int(0);
+  mp_integer constants=0;
+
+  for(auto it : m)
+  {
+    // We should group constants together...
+    const exprt &t=it.first;
+    int second=negated?(-it.second):it.second;
+    if(t.id()==ID_constant)
+    {
+      std::string value(to_constant_expr(t).get_value().c_str());
+      constants+=binary2integer(value, true)*second;
+    }
+    else
+    {
+      if(second!=0)
+      {
+        if(second==-1)
+        {
+          if(sum==refined_string_typet::index_of_int(0))
+            sum=unary_minus_exprt(t);
+          else
+            sum=minus_exprt(sum, t);
+        }
+        else if(second==1)
+        {
+          if(sum==refined_string_typet::index_of_int(0))
+            sum=t;
+          else
+            sum=plus_exprt(sum, t);
+        }
+      }
+      else
+      {
+        debug() << "in string_refinementt::sum_of_map:"
+                << " warning: several occurences of the same variable: "
+                << t.pretty() << eom;
+        variable_with_multiple_occurence_in_index=true;
+        if(second>1)
+        {
+          for(int i=0; i<second; i++)
+            sum=plus_exprt(sum, t);
+        }
+        else
+        {
+          for(int i=0; i>second; i--)
+            sum=minus_exprt(sum, t);
+        }
+      }
+    }
+  }
+
+  exprt index_const=from_integer(constants, refined_string_typet::index_type());
+  return plus_exprt(sum, index_const);
+}
+
+exprt string_refinementt::simplify_sum(const exprt &f)
+{
+  std::map<exprt, int> map=map_of_sum(f);
+  return sum_of_map(map);
+}
+
+exprt string_refinementt::compute_subst
```

Comment:
What does this do? Subst...itution, I guess? But from what, to what? Looking below, I guess this could be renamed "instantiate_quanfitier" or "substitute_quantifier" or similar? What is `f` for?